pkg/internal/apply/applier_integration_test.go

```diff
@@ -0,0 +1,188 @@
+	corev1 "k8s.io/api/core/v1"
+
+	//metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	k8sscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func combinedScheme() *runtime.Scheme {
+	intScheme := runtime.NewScheme()
+	_ = k8sscheme.AddToScheme(intScheme)   // nolint:errcheck // ok
+	_ = kraanscheme.AddToScheme(intScheme) // nolint:errcheck // ok
+	_ = hrscheme.AddToScheme(intScheme)    // nolint:errcheck // ok
+	return intScheme
+}
+
+func kubeConfigFromFile(t *testing.T) (*rest.Config, error) {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if len(kubeconfig) == 0 {
+		kubeconfig = filepath.Join(os.Getenv("HOME"), ".kube", "config")
+	} else {
+		t.Logf("Using KUBECONFIG at '%s'", kubeconfig)
+	}
+	t.Logf("Checking for KUBECONFIG file '%s'", kubeconfig)
+	info, err := os.Stat(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+	if !info.Mode().IsRegular() {
+		return nil, fmt.Errorf("KUBECONFIG '%s' is not a configuration file", kubeconfig)
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create a Kubernetes Config from KUBECONFIG '%s'", kubeconfig)
+	}
+	return config, nil
+}
+
+func kubeConfig(t *testing.T) (*rest.Config, error) {
+	config, err := kubeConfigFromFile(t)
+	if err == nil {
+		return config, nil
+	}
+	t.Logf("No KUBECONFIG from file '%s' - using InClusterConfig", err)
+	return rest.InClusterConfig()
+}
+
+func runtimeClient(t *testing.T, scheme *runtime.Scheme) client.Client {
+	config, err := kubeConfig(t)
+	if err != nil {
+		t.Fatalf("kubernetes config error: %s", err)
+	}
+	client, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		t.Fatalf("Unable to create controller runtime client: %s", err)
+	}
+	return client
+}
+
+func kubeCoreClient(t *testing.T) *kubernetes.Clientset {
+	config, err := kubeConfig(t)
+	if err != nil {
+		t.Fatalf("kubernetes config error: %s", err)
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("Unable to create Kubernetes API Clientset: %s", err)
+	}
+	return clientset
+}
+
+func TestConnectToCluster(t *testing.T) {
+	mockCtl := gomock.NewController(t)
+	defer mockCtl.Finish()
+
+	scheme := combinedScheme()
+	client := runtimeClient(t, scheme)
+	k8sclient := kubeCoreClient(t)
+
+	info, err := k8sclient.ServerVersion()
+	if err != nil {
+		t.Fatalf("Error getting server version: %s", err)
+	}
+	t.Logf("Server version %s", info.GitVersion)
+
+	namespaceList := &corev1.NamespaceList{}
+	err = client.List(context.Background(), namespaceList)
+	if err != nil {
+		t.Fatalf("runtime error getting namespaces: %s", err)
+	}
+	for _, namespace := range namespaceList.Items {
+		t.Logf("Found Namespace '%s'", namespace.GetName())
+	}
+
+}
+
+func TestSimpleApply(t *testing.T) {
```

Comment:
For the Test*Apply i wonder if you could combine into a table driven test instead?
https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Reply:
I _could_ but I didn't.  Right now, I'm using this test for isolated functional testing rather than actual automated integration testing.  We might eventually convert most of our tests into table-driven tests something like Dave Cheney's pattern, but we're not doing TDD at this stage.